backend/icloud/icloud.go

```diff
@@ -0,0 +1,1149 @@
+	dstDirectoryID, _ := f.parseNormalizedID(jdstDirectoryID)
+
+	_, err = f.move(ctx, srcID, srcDirectoryID, srcLeaf, srcEtag, dstDirectoryID, dstLeaf)
+	if err != nil {
+		return err
+	}
+
+	srcFs.dirCache.FlushDir(srcRemote)
+
+	return nil
+}
+
+func (f *Fs) move(ctx context.Context, ID, srcDirectoryID, srcLeaf, srcEtag, dstDirectoryID, dstLeaf string) (*api.DriveItem, error) {
+	var resp *http.Response
+	var item *api.DriveItem
+	var err error
+
+	fs.Debugf(ID, "ID")
+	fs.Debugf(srcDirectoryID, "srcDirectoryID")
+	fs.Debugf(srcLeaf, "srcLeaf")
+
+	// move
+	if srcDirectoryID != dstDirectoryID {
+		if err = f.pacer.Call(func() (bool, error) {
+			id, _ := f.parseNormalizedID(ID)
+			item, resp, err = f.service.MoveItemByItemID(ctx, id, srcEtag, dstDirectoryID, true)
+			return ignoreResultUnknown(ctx, resp, err)
+		}); err != nil {
+			return nil, err
+		}
+		ID = item.Docwsid
+		srcEtag = item.Etag
+	}
+
+	// rename
+	if srcLeaf != dstLeaf {
+		if err = f.pacer.Call(func() (bool, error) {
+			id, _ := f.parseNormalizedID(ID)
+			item, resp, err = f.service.RenameItemByItemID(ctx, id, srcEtag, dstLeaf, true)
+			return ignoreResultUnknown(ctx, resp, err)
+		}); err != nil {
+			return item, err
+		}
+	}
+
+	return item, err
+}
+
+// Move moves the src object to the specified remote.
+func (f *Fs) Move(ctx context.Context, src fs.Object, remote string) (fs.Object, error) {
+	srcObj, ok := src.(*Object)
+	if !ok {
+		fs.Debugf(src, "Can't move - not same remote type")
+		return nil, fs.ErrorCantMove
+	}
+
+	srcLeaf, srcDirectoryID, _, err := srcObj.fs.FindPath(ctx, srcObj.remote, true)
+	if err != nil {
+		return nil, err
+	}
+
+	dstLeaf, dstDirectoryID, _, err := f.FindPath(ctx, remote, true)
+	if err != nil {
+		return nil, err
+	}
+
+	item, err := f.move(ctx, srcObj.docID, srcDirectoryID, srcLeaf, srcObj.etag, dstDirectoryID, dstLeaf)
+	if err != nil {
+		return src, err
+	}
+
+	return f.NewObjectFromDriveItem(ctx, remote, item)
+}
+
+// Creates from the parameters passed in a half finished Object which
+// must have setMetaData called on it
+//
+// Returns the object, leaf, directoryID and error.
+//
+// Used to create new objects
+func (f *Fs) createObject(ctx context.Context, remote string, modTime time.Time, size int64) (o *Object, err error) {
+	// Create the directory for the object if it doesn't exist
+	_, _, _, err = f.FindPath(ctx, remote, true)
+	if err != nil {
+		return
+	}
+	// Temporary Object under construction
+	o = &Object{
+		fs:      f,
+		remote:  remote,
+		modTime: modTime,
+		size:    size,
+	}
+	return o, nil
+}
+
+// ReadCookies parses the raw cookie string and returns an array of http.Cookie objects.
+func ReadCookies(raw string) []*http.Cookie {
+	header := http.Header{}
+	header.Add("Cookie", raw)
+	request := http.Request{Header: header}
+	return request.Cookies()
+}
+
+var retryErrorCodes = []int{
+	400, // icloud is a mess, sometimes returns 400 on a perfectly fine request. So just retry
```

Comment:
Eww that is untidy! Better hope we don't make any mistakes with parameters then!

Reply:
i know but its rampant, unit test wint even succeed without it :( Theres not really a way for me to know if its a mail formed request or an icloud issue otherwise i would have added better validation.